### PR TITLE
Implement float/vec operations in RISC-V jit

### PR DESCRIPTION
**Pull request by unknownbrackets**
This improves performance a good bit for games that were still slow.

LittleBigPlanet FPS doubles.  Fat Princess improves by 150% (although still under 100% on a VisionFive 2.)  This is mostly due to heavy float usage.

Even features a commit from like 2016 to make rounding mode changes actually work.  With this, only the flush-to-zero test fails in the cpu/fpu/fpu test (meaning it passes more than IR interp, which might be fixable using some intrinsic probably.)

-[Unknown]

### CODE REVIEW

File: Core/MIPS/ARM64/Arm64RegCacheFPU.cpp [hunk 1]
```diff
@@ -319,6 +319,7 @@ void Arm64RegCacheFPU::FlushR(MIPSReg r) {
 		if (mr[r].reg == INVALID_REG) {
 			ERROR_LOG(JIT, "FlushR: MipsReg had bad ArmReg");
 		}
+		FlushArmReg((ARM64Reg)(S0 + mr[r].reg));
```
Comment (hrydgard) [inline Core/MIPS/ARM64/Arm64RegCacheFPU.cpp:322]: This Arm64 change maybe belongs to a different PR? Though not saying it looks wrong. Did you notice a problem?

File: Core/MIPS/RiscV/RiscVCompFPU.cpp [hunk 2]
```diff
@@ -109,8 +109,28 @@ void RiscVJit::CompIR_FAssign(IRInst inst) {
 		break;
 
 	case IROp::FSign:
-		CompIR_Generic(inst);
+	{
+		fpr.MapDirtyIn(inst.dest, inst.src1);
+		// Check if it's negative zero, either 0x10/0x08 is zero.
+		FCLASS(32, SCRATCH1, fpr.R(inst.src1));
```
Comment (hrydgard) [inline Core/MIPS/RiscV/RiscVCompFPU.cpp:143]: huh, interesting instruction! And I guess likely to be faster than moving the fp reg to a GPR and implementing this through bit operations on that.

File: Core/MIPS/RiscV/RiscVCompFPU.cpp [hunk 3]
```diff
@@ -51,7 +51,35 @@ void RiscVJit::CompIR_FArith(IRInst inst) {
 	case IROp::FMul:
 		fpr.MapDirtyInIn(inst.dest, inst.src1, inst.src2);
 		if (inst.src1 != inst.src2) {
-			CompIR_Generic(inst);
+			// These will output 0x80/0x01 if infinity, 0x10/0x80 if zero.
```
Comment (hrydgard) [inline Core/MIPS/RiscV/RiscVCompFPU.cpp:54]: Hm, on ARM64 we just emit an FMUL instruction, since it's behavior in inf*0 is the same as for MIPS, [unlike x86](https://github.com/unknownbrackets/ppsspp/commit/5afc020fbdbf6e9f9139fbc4919430af3a652285). I guess you noticed through tests that RISC-V is closer to x86 here?

File: Core/MIPS/RiscV/RiscVCompVec.cpp [hunk 4]
```diff
@@ -189,7 +189,39 @@ void RiscVJit::CompIR_VecHoriz(IRInst inst) {
 
 	switch (inst.op) {
 	case IROp::Vec4Dot:
-		CompIR_Generic(inst);
+		// TODO: Maybe some option to call the slow accurate mode?
```
Comment (hrydgard) [inline Core/MIPS/RiscV/RiscVCompVec.cpp:192]: I think we can just add a check here for the compat.ini option?

**Review by hrydgard — APPROVED** [review]

File: Core/MIPS/ARM64/Arm64RegCacheFPU.cpp (hunk 1 above)
Comment (unknownbrackets) [reply Core/MIPS/ARM64/Arm64RegCacheFPU.cpp:322]: Well, nothing is actually using FlushR() or FlushV() in arm64jit currently, so this actually changes nothing.  But FlushR() was accidentally a DiscardR() before, because it didn't actually flush it, so I figured I'd fix it in case we ever use it...

-[Unknown]

File: Core/MIPS/RiscV/RiscVCompFPU.cpp (hunk 3 above)
Comment (unknownbrackets) [reply Core/MIPS/RiscV/RiscVCompFPU.cpp:54]: Well, I was worried because it doesn't specify much and there's different people implementing RISC-V.  Maybe we can remove, but I'd want to have more confidence in it working on more chips first... I didn't actually verify if FMUL was doing this or not.

-[Unknown]

File: Core/MIPS/RiscV/RiscVCompVec.cpp (hunk 4 above)
Comment (unknownbrackets) [reply Core/MIPS/RiscV/RiscVCompVec.cpp:192]: Well, that option just does vmmul, I think.  It may be useful at some point to have a slower-but-accurate-vdot mode to check if it's the cause of bugs, without having to go all the way to interpreter for most of VFPU.  But I dunno, maybe it's fine.  Was just thinking about it when I wrote Vec4Dot...

-[Unknown]

File: Core/MIPS/RiscV/RiscVCompFPU.cpp (hunk 3 above)
Comment (hrydgard) [reply Core/MIPS/RiscV/RiscVCompFPU.cpp:54]: Okay, that's reasonable. Please add that note as a code comment though :)

File: Core/MIPS/RiscV/RiscVCompVec.cpp (hunk 4 above)
Comment (hrydgard) [reply Core/MIPS/RiscV/RiscVCompVec.cpp:192]: Hm, I think we use it for a few more ops too? Though yeah, we'd need to track where the Vec4Dot came from, which seems silly. We would rather have two different Vec4Dot opcodes I think, one like this and one more accurate, and emit according to the compat option.

**Comment by archanox** [comment]
Tested successfully on a th1520 (c910 cores) found in a LicheePi4A

**Comment by archanox** [comment]
Tested successfully on a th1520 (c910 cores) found in a LicheePi4A
